src/data/indicator_function.lean

```diff
@@ -0,0 +1,116 @@
+
+lemma indicator_congr (h : ∀ a ∈ s, f a = g a) : indicator s f = indicator s g :=
+funext $ λx, by { simp only [indicator], split_ifs, { exact h _ h_1 }, refl }
+
+@[simp] lemma indicator_univ (f : α → β) : indicator (univ : set α) f = f :=
+funext $ λx, indicator_of_mem (mem_univ _) f
+
+@[simp] lemma indicator_empty (f : α → β) : indicator (∅ : set α) f = λa, 0 :=
+funext $ λx, indicator_of_not_mem (not_mem_empty _) f
+
+variable (β)
+@[simp] lemma indicator_zero (s : set α) : indicator s (λx, (0:β)) = λx, (0:β) :=
+funext $ λx, by { simp only [indicator], split_ifs, refl, refl }
+variable {β}
+
+lemma indicator_indicator (s t : set α) (f : α → β) : indicator s (indicator t f) = indicator (s ∩ t) f :=
+funext $ λx, by { simp only [indicator], split_ifs, repeat {simp * at * {contextual := tt}} }
+
+lemma indicator_preimage (s : set α) (f : α → β) (B : set β) :
+  (indicator s f)⁻¹' B = s ∩ f ⁻¹' B ∪ (-s) ∩ (λa:α, (0:β)) ⁻¹' B :=
+by { rw [indicator, if_preimage] }
+
+end has_zero
+
+section has_add
+variables [add_monoid β] {s t : set α} {f g : α → β} {a : α}
+
+lemma indicator_union_of_not_mem_inter (h : a ∉ s ∩ t) (f : α → β) :
+  indicator (s ∪ t) f a = indicator s f a + indicator t f a :=
+by { simp only [indicator], split_ifs, repeat {simp * at * {contextual := tt}} }
+
+lemma indicator_union_of_disjoint (h : disjoint s t) (f : α → β) :
+  indicator (s ∪ t) f = λa, indicator s f a + indicator t f a :=
+funext $ λa, indicator_union_of_not_mem_inter
+  (by { convert not_mem_empty a, have := disjoint.eq_bot h, assumption })
+  _
+
+lemma indicator_add (s : set α) (f g : α → β) :
+  indicator s (λa, f a + g a) = λa, indicator s f a + indicator s g a :=
+by { funext, simp only [indicator], split_ifs, { refl }, rw add_zero }
+
+lemma indicator_smul {𝕜 : Type*} [monoid 𝕜] [distrib_mul_action 𝕜 β] (s : set α) (r : 𝕜) (f : α → β) :
+  indicator s (λ (x : α), r • f x) = λ (x : α), r • indicator s f x :=
+by { simp only [indicator], funext, split_ifs, refl, exact (smul_zero r).symm }
+
+lemma indicator_neg {β : Type*} [add_group β] (s : set α) (f : α → β) :
+  indicator s (λa, - f a) = λa, - indicator s f a :=
+by { funext, simp only [indicator], split_ifs, { refl }, rw neg_zero }
+
+lemma indicator_sub {β : Type*} [add_group β] (s : set α) (f g : α → β) :
+  indicator s (λa, f a - g a) = λa, indicator s f a - indicator s g a :=
+by { funext, simp only [indicator], split_ifs, { refl }, rw sub_zero }
+
+end has_add
+
+section order
+variables [has_zero β] [preorder β] {s t : set α} {f g : α → β} {a : α}
+
+lemma indicator_le_indicator (h : f a ≤ g a) : indicator s f a ≤ indicator s g a :=
+by { simp only [indicator], split_ifs with ha, { exact h }, refl }
+
+lemma indicator_le_indicator_of_subset (h : s ⊆ t) (hf : ∀a, 0 ≤ f a) (a : α) :
+  indicator s f a ≤ indicator t f a :=
+begin
+  simp only [indicator],
+  split_ifs,
+  { refl },
+  { have := h h_1, contradiction },
```

Comment:
I'm not sure if relying on an autogenerated name `h_1` is a good idea.

Reply:
Mainly because `split_ifs` is bugged. If there are four cases, it ignores the names I provided on the third and fourth case.

For example, 
```lean
lemma indicator_congr_of_set [measure_space α] (h : ∀ₘ a, a ∈ s ↔ a ∈ t) :
  ∀ₘ a, indicator s f a = indicator t f a :=
begin
  filter_upwards [h],
  simp only [mem_set_of_eq, indicator],
  assume a ha,
  split_ifs with h₁ h₂,
  { refl },
  { have := ha.1 h₁, contradiction },
  { have := ha.2 h_1, contradiction }, -- gives me h_1 instead of h₂
  refl                                                     -- same thing as above
end
```

Reply:
It seems that `split_ifs with h₁ h₂ h₂` works. I think that it unfolds the first `if`, then looks for remaining `if`s in each branch, and fails to identify "the same" `if`. E.g., `if p then if q then a else b else c` results in 3 branches.

Reply:
That's true! Thanks.